src/hash.rs

```diff
@@ -64,3 +84,80 @@ fn file_hash(algorithm: &str, path: &str) -> Result<String> {
 
     Ok(hash_algorithm(algorithm, &bytes)?)
 }
+
+/// Generates multiple TOTP codes from 20 character hex_seed, with time step +-tolerance
+/// time_override is used as the current unix time instead of the current system time for testing
+fn totp_generate_tolerance(hex_seed: &str, tolerance: i32, time_override: Option<i64>) -> Result<String> {
+    let mut results: Vec<String> = Vec::new();
+    for i in -tolerance..(tolerance + 1) {
+        let result = totp_generate(hex_seed, i.try_into().unwrap(), time_override)?;
+        if result.starts_with("ERROR:") {
```

Comment:
You definitely want code like this to be closer to:

```rust
let result = totp_generate(hex_seed, i.try_into().unwrap(), time_override)?
results.push(result);
```

Which I assume is what you were actually going for before being blocked by the functions' Result type.

Reply:
The problem is I couldn't return a `FromHexError` and when I tried to instantiate `Error` I got an error I didn't understand, so I'm just returning an Ok, with the error message as the text.

Reply:
If I instead try to return `Err(String)` I get 'expected enum `error::Error`' and if I try `Err(Error::new)` I get 'variant or associated item cannot be called on `error::Error` due to unsatisfied trait bounds'

Reply:
With `Error::new` I assume you were trying to create an instance of the Error enum in our error.rs module? It's an enum so to create an instance you can pick one of the variants and use it directly.

For example our first variant is Null: https://github.com/tgstation/rust-g/blob/539c948a5132c3617045026f2fedfaeb23d5c99a/src/error.rs#L19-L22

To use that in some code you could do:
```rust
fn my_func() -> Result<u32> {
    return Err(Error::Null)
}
```

Reply:
Can I add my own error to the Enum? None of them fit very well

Reply:
I now see that is what you suggested in the original comment